experiments/kompanion/cmd/migrator/migrator.go

```diff
@@ -0,0 +1,442 @@
+	}
+
+	// write the namespaces to the migrator file
+	filedir := filepath.Dir(opts.migrationFile)
+	if err := os.MkdirAll(filedir, 0o700); err != nil {
+		return fmt.Errorf("error creating directory %s: %w", filedir, err)
+	}
+	migfile, err := os.Create(opts.migrationFile)
+	if err != nil {
+		return fmt.Errorf("error creating migration file %s: %w", opts.migrationFile, err)
+	}
+	defer migfile.Close()
+	nsGvr := schema.GroupVersionResource{
+		Group:    "",
+		Version:  "v1",
+		Resource: "namespaces",
+	}
+	nss, err := dynamicClient.Resource(nsGvr).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("fetching namespace %s resources: %w", nsGvr, err)
+	}
+	for _, ns := range nss.Items {
+		ns = pruneResource(ns)
+		if len(ns.GetAnnotations()) <= 0 {
+			continue
+		}
+		data, err := yaml.Marshal(ns)
+		if err != nil {
+			return fmt.Errorf("error marshalling namespace: %w", err)
+		}
+		migfile.WriteString(string(data[:]))
+		migfile.WriteString("\n---\n")
+	}
+
+	// create the work log for go routine workers to use
+	q := &taskQueue{}
+
+	// Parallize across resources, unless we are scoped to a few namespaces
+	// The thought is that if users target a particular namespace (or a few), they may not have cluster-wide permission.
+	perNamespace := len(opts.targetNamespaces) > 0
+	if perNamespace {
+		for _, ns := range namespaces.Items {
+			if shouldExclude(ns.Name, opts.ignoreNamespaces, opts.targetNamespaces) {
+				continue
+			}
+
+			q.AddTask(&dumpResourcesTask{
+				Namespace:     ns.Name,
+				Resources:     resources,
+				DynamicClient: dynamicClient,
+				OutputChannel: outputChannel,
+				// ReportDir:           reportTempDir,
+				ShouldExcludeObject: shouldExcludeObject,
+			})
+		}
+	} else {
+		for _, resource := range resources {
+			q.AddTask(&dumpResourcesTask{
+				Resources:     []schema.GroupVersionResource{resource},
+				DynamicClient: dynamicClient,
+				OutputChannel: outputChannel,
+				// ReportDir:           reportTempDir,
+				ShouldExcludeObject: shouldExcludeObject,
+			})
+		}
+	}
+
+	var wg sync.WaitGroup
+
+	var errs []error
+	var errsMutex sync.Mutex
+
+	for i := 0; i < opts.workerRoutines; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			for {
+				task := q.GetWork()
+				if task == nil {
+					// no work
+					return
+				}
+
+				if err := task.Run(ctx); err != nil {
+					errsMutex.Lock()
+					errs = append(errs, err)
+					errsMutex.Unlock()
+				}
+			}
+		}()
+	}
+
+	go func() {
+		defer wg.Done()
+
+		for data := range outputChannel {
+			migfile.WriteString(data)
+			migfile.WriteString("\n---\n")
+		}
+	}()
+
+	wg.Wait()
+	close(outputChannel)
+	return nil
+}
+
+func resourcesToName(resources []*metav1.APIResource) []string {
+	names := make([]string, len(resources))
+
+	for i, resource := range resources {
+		names[i] = resource.SingularName
+	}
+
+	return names
+}
+
+func shouldExclude(name string, excludes []string, includes []string) bool {
+	// todo acpana: maps
```

Comment:
Or sets :-)

Reply:
Future fix. Removed the comment for now.